### ISSUE TITLE
Better documentation for tergm control settings

### ISSUE THREAD

**Issue by martinamorris** [EpiModel/EpiModel#743]
The documentation on the control settings leaves much to be desired. This could be improved with a vignette or separate standalone help page in the package.

Here is the tortuous current path:  

The `set.control.tergm` function can't be queried directly:

`?EpiModel::set.control.tergm` and the `??` version both come up empty

`?EpiModel::netest` points to `tergm::simulate.network` and

`tergm::simulate.network` points to `tergm::simulate.tergm`

**Comment by smjenness** [comment]
Sounds great, and would be nice to have something new for SISMID this year.

**Comment by smjenness** [comment]
I am going to bump this to 2.7.0 so I can get this version of EpiModel up to CRAN, but we can still develop docs on SISMID for the course.

**Comment by sgoodreau** [comment]
OK I've begun working on this.  Ultimately I believe we want material in three places:
- a vignette
- man pages (e.g. `?control/tergm`)
- SISMID teaching materials

I'm starting with the vignette since it's nice and stand-alone and allows me to explore examples easily in `.qmd` format until I get them fine-tuned. Once I make some good progress there we can discuss about the most important content to port over to the other places.

The reason exploration is needed is that I'm going into this semi-blind, in that I don't feel like I actually have a good sense of what all the best options are, and when which is most useful. The sources I'm going to start with to get control parameter candidates to experiment with are laid out below. @smjenness @martinamorris can you think of any other big ones I'm missing?  I'd like to hold off on bothering Pavel until I've made as much progress as I can with existing materials.

`tergm` control parameters:
- Search on prior discussions about the best control parameters to tweak in different circumstances (I know these have involved Pavel, Carter, Chad, Aditya and Michal at least, and I even compiled these in a document somewhere, which I didn’t find on a shallow dive but will do a deep dive now, across email, git, and files).
- ?tergm / ?control.tergm
- The main tergm paper (Krivitsky and Handcock 2014)
- Sunbelt workshop (https://statnet.org/workshop-tergm/)

Along the way I imagine it will help to refresh my knowledge of `ergm` control parameters, with these sources:
- ?ergm / ?control.ergm
- ERGM 4.0 paper (Krivitsky et al. 2023)	
- Sunbelt workshop (https://statnet.org/workshop-ergm/)

**Comment by smjenness** [comment]
@sgoodreau Great plan. Two thoughts: one on sequencing, one on sources that may be helpful to consider.

**Sequencing:** keep authoring in `.qmd` like you're doing, but I suggest you aim to put the polished version on the SISMID materials *first*, and treat the package vignette + man pages for the next CRAN release. It's purely about reach: a packaged vignette doesn't get to users until v2.7.0 is on CRAN (published after SISMID), whereas course/website materials we can publish and revise right away.

**Big sources for your consideration:** the most useful material related to this is in our own production code and issue history. In practice we only ever touch a small subset of these controls, so:

1. **[EpiModelHIV-Template](https://github.com/EpiModel/EpiModelHIV-Template) estimation code** (`R/A-networks/z-context.R` plus the `model_main/casl/ooff.R` files). This is the de facto "what we actually set" reference. The shared `control.ergm` object there basically *is* the shortlist: `main.method` (we flip between `"MCMLE"` on HPC and `"Stochastic-Approximation"` locally), `MCMLE.maxit = 500`, `MCMC.samplesize = 1e4`, `MCMC.interval = 5e3`, `SAN.maxit` / `SAN.nsteps.times`, and `parallel`. Those five or six are ~95% of what we change when an egocentric fit misbehaves.

2. **Our own closed issues** are the most helpful archive of Chad's tuning advice (perhaps more findable than the doc you compiled). The high-value ones:
   - **#435**: Chad's complete large + constrained recipe, plus the key intuition that shorter duration at fixed edges means more dyad changes per step, so you raise `MCMC.burnin.min`/`.max`. Best single read.
   - **#522**: the canonical "big network won't hit target stats" walkthrough, with `MCMC.init.maxchanges` for under-fit edge counts and a clean side-by-side of estimation vs. dynamic-sim controls.
   - **#779**: the authoritative "tergm has no `MCMC.interval`" explanation (set `MCMC.burnin.min = MCMC.burnin.max` to half the desired proposals/step), plus the real tergm mixing tools (`discordance_fraction`, `~strat(...) + discord + sparse`).
   - **#1026**: why "dynamic looks great, static looks terrible" diagnostics are an effective-sample-size / `MCMC.interval` artifact rather than a convergence bug. Directly relevant to your textbook section, and a good "don't over-tune" counterweight.
   - **#519** and **#438**: the boundary where tuning *won't* help. Version-to-version netdx differences are driven by the EDA formula change, not MCMC controls (#519), and very short durations are an approximation problem that needs a full STERGM, not more burnin (#438).
   - **#543** / **#673**: the conceptual taxonomy of why there are so many control objects at all (separable stergm vs. non-separable tergm, fitted-object vs. formula simulation, tergmLite needing every control present).

3. **Scoping suggestion:** given the above, I think the vignette is most helpful if it's organized around the ~8 parameters we typically use, split by stage, rather than the full `control.ergm` / `control.tergm` set:
   - *Estimation* (`control.ergm`): `main.method`, `MCMLE.maxit`, `MCMC.samplesize`, `MCMC.interval`, `MCMC.burnin`, `SAN.*`, `init.MPLE.samplesize`, `parallel`.
   - *Dynamic simulation* (`control.simulate.network` / `set.control.tergm`): `MCMC.burnin.min`/`.max`, `MCMC.init.maxchanges`, and the `MCMC.prop` / `discordance_fraction` mixing knobs.

**Comment by sgoodreau** [comment]
@smjenness woo-hoo you just saved me days of work :-)

As for aiming for SISMID first: no problem there. Do you want me to integrate it into an existing tutorial now, or keep it as its own thing and we'll figure out later where it fits in?

**Comment by smjenness** [comment]
Great! I would format as a standalone tutorial and we can figure out later how to integrate it.

**Comment by smjenness** [comment]
Following up on my earlier note that our production code and issue history are the best source of "what we actually set" cases. We just hit a clean one fitting the GlobalMix India contact networks, and it exercises almost exactly the short list of controls we said are ~95% of what we touch. Posting it here as a candidate case study for the vignette / SISMID materials, because the same model fits on one network and stalls for a day on another, which makes a nice teachable contrast.

## The setup

24 dynamic networks: 2 settings (rural, urban) x 3 layers (school, work, non-home) x 4 population fractions (0.25, 0.5, 0.75, 1.0), each fit independently. Sizes run from N ~ 29k to ~258k. Each layer is a dyad-independent cross-sectional formation model

```
~edges + nodemix("age.grp", levels2 = -ref) + nodefactor("deg_<other layer>", levels = -1)
```

with an EDA dissolution (`~offset(edges)`, duration from data). The target stats are scaled per-capita counts, so they are fractional: SAN can never match them exactly, so `netest()` always falls back to MCMLE even though the model is dyad-independent. Several elderly age-mixing cells are structural zeros (no school contacts for 60+, etc.).

## The failure

With a control that looks reasonable (`MCMC.samplesize = 5e4`, default full step, no `MCMC.maxedges`):

```
Iteration 1 of at most 200:
1 Optimizing with step length 1.0000.
The log-likelihood improved by 4.17.
Estimating equations are not within tolerance region.
Iteration 2 of at most 200:
   <16 cores pegged at 98% for 23h, no further output>
```

Rural school and the dense non-home layers did this; the *identical* urban school model converged in ~9 iterations / 2.5h. So it is not raw N (urban N=258k > rural N=118k), it is the structure: after iteration 1 takes a full step into a slow-mixing region, the large fixed MCMC sample (5e4 x 5e3 ~ 2.5e8 proposals) makes a single draw there enormous, and the default density guard (~20x observed edges) is loose enough on a network this dense that the chain mixes slowly under it rather than aborting. The default `MCMLE.termination = "confidence"` would compound this across iterations by boosting the sample size, though here the run never even cleared iteration 2.

## What actually fixed it (the short list)

We ran a controlled sweep, one strategy per job. Several worked; they all attack the same two things, *bound the per-iteration cost* and *stop iteration 1 from overshooting*:

| control | what it did | note |
|---|---|---|
| `MCMC.samplesize = 1e4` (down from 5e4) + `MCMC.maxedges = 5e6` | the single biggest lever: a small fixed sample makes every iteration cheap, and a tight finite `maxedges` (well below the default ~20x density-guard ceiling) turns a degenerate dense draw into a fast status-1 error instead of a slow draw | keep `MCMC.effectiveSize = NULL` so the within-iteration sample stays bounded |
| `MCMLE.steplength = 0.5` **with** `MCMLE.steplength.margin = NULL` | pins the step so iteration 1 cannot lunge into the bad region | the `margin = NULL` is load-bearing and non-obvious: with the default margin, `steplength` is treated only as the maximum and ergm's Hummel search raises the actual step up toward it, so setting `steplength` alone does not pin the step |
| `MCMLE.termination = "Hummel"` | terminates without the "confidence" between-iteration sample blow-up | requires `steplength = 1` + non-null margin, so it cannot be combined with a pinned step (ergm errors if you try) |
| `MCMC.prop = ~strat(attr = ~age.grp) + sparse` | stratified proposal so the sampler stops starving the rare elderly cells; fastest on the *dense* non-home layer | the estimation-side use of the proposal knob, not just simulation |
| `SAN.maxit = 12, SAN.nsteps.times = 16` | strong SAN so the MPLE exists | weak SAN gave "The MPLE does not exist!" on the structural-zero cells |

Net effect on the worst case (rural school, N=118k): from a 23h non-convergence to ~2.4 min with the small-sample + Hummel recipe. The whole 24-model grid now re-fits in about an hour, and a dynamic `netdx` on all 24 confirms it: edges recover to within 1.2% on every layer (most under 0.5%), with mean durations and dissolution within 0.25%. The layer character decides the best proposal: small-sample + Hummel is fastest on the sparse school/work layers, the stratified proposal is fastest and cleanest on the dense non-home layers.

One caveat for the docs: the specific magnitudes here are tuned to these networks. The transferable content is which controls matter and why, not the literal values.

## Two diagnosis traps worth teaching

These cost us time and I think both belong in the docs, because the "obvious" reading is wrong in each:

1. **Do not blame the structural-zero cells.** ergm 4.x already detects them ("Observed statistic(s) ... are at their smallest attainable values. Their coefficients will be fixed at -Inf.") and fixes them. They are not the direct cause of the stall; the slow mixing is among the *remaining* free parameters after the overshoot. Adding an explicit constraint for them did not help.

2. **"Static recovery looks terrible" was a burn-in artifact, not a bad fit.** When we sanity-checked a fit by simulating from an empty network and comparing to targets, edges came back far short of target and it looked like a broken fit. It was the effective-sample / burn-in issue: from an empty start the default burn-in is nowhere near equilibrium, and the gap widens with N (at the full network the default-burn-in recovery was ~0.05). A controlled burn-in sweep on one layer shows it is a gradient, not a cliff, and that the same coefficients recover fine once the chain has time to fill in:

| `simulate` MCMC.burnin | edges recovered |
|---|---|
| default | 0.13 |
| 1e6 | 0.79 |
| 1e7 | 0.99 |

This is the same point as #1026, and it is easy to mistake for a convergence failure when it is really a diagnostics-side setting.

## The scale-up shortcut (ties directly to the SISMID open-pop module)

Because the formation model is dyad-independent, we can skip the large-N fit entirely. Fit the layer at pct=0.1 (small, fast, stable with Stochastic-Approximation), then move the coefficients to the full network unchanged *except* edges, which shifts by `log(N_small) - log(N_big)` to hold mean degree fixed (exactly the open-pop edges adjustment in the SISMID materials):

```r
theta <- coef(fit_small$fit)
theta["edges"] <- theta["edges"] + log(N_small) - log(N_big)   # nodemix/nodefactor unchanged
```

Simulating the big network from these transferred coefficients (with an adequate burn-in, per the trap above) recovers the targets to ~0.99, and an independent full MCMLE fit lands on the same edges coefficient to three decimals. Used as the MCMLE `init`, it also converges cleanly. This is a tidy demonstration that, in the sparse regime with per-capita targets, the dyad-independent differential terms are approximately invariant to N and the size dependence is carried almost entirely by the baseline edges term. It rests on the sparse `plogis ~ exp` approximation, which is why it recovers to ~0.99 rather than exactly.

## What did not work (useful as counterexamples)

- Plain `control.ergm()` defaults: degenerate, edges blew up >20x in 85s.
- Stochastic-Approximation at full N: fast but over/under-recovers (edges 1.09 with one cell 9x off at samplesize 1e5; 3.7x over at 5e5, i.e. a *bigger* SA sample was worse here).
- Warm-starting from the converged urban fit's `coef.form`: degenerate, because `coef.form` is the dissolution-adjusted formation coefficient, not the cross-sectional MLE, so it is the wrong scale for an `init`. The same-setting small-N transfer above is the clean version.

Happy to package the harness (one strategy per SLURM array task) and the per-layer logs if useful for the vignette. The controls above map onto the estimation shortlist you sketched earlier in this thread, with `MCMC.maxedges`, `MCMLE.steplength.margin`, and the `~strat() + sparse` proposal being the three that I would not have guessed mattered as much as they did.

**Comment by sgoodreau** [comment]
@smjenness @martinamorris I wanted to get clarification around my task here.  The title of the issue is "Better documentation for tergm control settings".  In my head, then, I was to be focusing on cases where a fit with `edapprox = TRUE` is poor, and one needs to run a full STERGM.

And yet, among the "Big sources for your consideration" Sam shared above, folks seem to be using the approximation, and the discussion is about `ergm` control settings, not `tergm`. (Not all; one exception is the pair `MCMC.burnin.min=2e7` amd `MCMC.burnin.max=2e7`.) But AFAIK that wouldn't help with a case where the approximation is just destined to be insufficient (e.g. with short relational durations), and I thought it was that latter case I was working on.

So: am I minsuderstanding my task? Or misunderstanding how helpful specifically `ergm` parameters can be in improving the `tergm` fits, since they still get used at some point? Or something else?

To be clear, it's been a very very long time since I've actually fit a research model with `edapprox = FALSE`. Actually it's been a while since I fit a research-level model at all.

**Comment by smjenness** [comment]
@sgoodreau The issue title is doing us a disservice. "tergm control settings" is Martina's original framing, but the work I think we actually need is broader: documenting the controls that flow through `netest` / `netdx` / `netsim` into both `ergm` and `tergm`, organized by the stage where they act.

Three stages to organize around:

1. Estimation (`control.ergm`, in `netest`). The cross-sectional ERGM fit, and where most of what we actually touch lives: `main.method`, `MCMLE.maxit`, `MCMC.samplesize`, `MCMC.interval`, `MCMC.burnin`, `MCMC.maxedges`, `MCMLE.steplength` + `steplength.margin`, `MCMLE.termination`, `SAN.*`, `init.MPLE.samplesize`, and the `strat` proposal. The India case (EpiModel/India_COVID19_model#38) sits entirely here, as does the estimation half of #435 and #522.

2. The dynamic-simulation refresh. Since we always run a fresh forward simulation from the fitted coefficients before stepping the epidemic, there are both the ergm simulation controls (set.control.ergm) and tergm simulation controls (set.control.tergm) in both netdx and netsim.

3. The boundary where no control helps. Short durations or high density break the ED approximation itself (#438; #519 is the related "the difference you're chasing is the EDA formula change, not your controls" case). This is the case you had in mind, but the practical wrinkle is just that at our network sizes a full STERGM doesn't scale, so what we actually reach for is rescaling the time step. I'm not sure what documentation we can provide at this stage.

**Comment by sgoodreau** [comment]
Thanks Sam, that all makes sense.  And I agree that documenting all of those collectively and coherently is something we should shoot for.  Not sure I can get on all of them before SISMID since it’s a larger lift than I understood from the title of the issue, but whatever we are able to add is better than what we already have.

That said, my reason for focusing on tergm fitting wasn’t only because it’s in the title. It’s actually been forefront in my mind for a bit -- my big concern here is actually on the book side, and I’m a fair bit worried, TBH. I get that in complex research models, there are lots of terms and large populations, and ergm itself struggles; a full stergm is pretty much out of the question in most cases. But in terms of pedagogy – as one starts out with a very simple model on a small population (as we do), and one has the chance to play around with it even a little bit (as is the case when someone has the luxury of time, as with a self-study book), then it becomes obvious very quickly that most models that aren’t carefully selected have issues with the approximation, and need a full stergm. You descibe this as a "boundary" case but it's actually most cases early on in the process of learning through exploration. I believe we have to be able to show folks here at least some options that meaningfully expand their range of fittable models, even if we ultimately say down the line that for a certain level of realistic complexity on larger population sizes, they’re just going to have to re-define their time step. We can’t just say “here’s all this great theory and machinery, but you can’t actually use any of it ever, even with toy examples”. If we do, most folks will just leave and never get to the later stages. 

In other words, I’m just fully realizing that we still don’t after all these years have much advice to offer people who need to do a full stergm, despite the existence of all those fitting parameters. That seems to be the case in the materials you pointed to, and it seems to be what you’re saying here.  I had had in the back of my mind all year that I needed to look into this, hoping that things had gotten better, and honestly kept putting it off of out nervousness and more timely things to do. If there really are very few options here, then I’m honestly not sure it even makes sense to continue with the book.

Of course you didn’t say there were no options here, just not many.  And maybe when you said that you were really thinking about the kind of models you work on regularly, not the relatively simple ones in SISMID or the book where we might actually have some ability to fit a stergm with a reasonable amount of parameter tweaking.

I spent a bit of time yesterday setting up some examples and just beginning to explore one set of fitting parameters (EGMME.MCMC.burnin.min, EGMME.MCMC.burnin.max). It’s slow going with a lot of trial and error, although I think I may see a bit of a glimmer of hope.  I’ll spend more time on it today.  Any advice you have in this area is highly highly (highly) welcome.

**Comment by smjenness** [comment]
I honestly have had no research-level needs for a full STERGM. Maybe something in the past for a day or two before I figured out some other solution. I would imagine it's easy to find clear exceptions when building new models to test and demonstrate the dimensions of our tools. So give me a good example or 10 here or over on the book's issues and I'll explore.

**Comment by sgoodreau** [comment]
The short answer as far as I can tell is pretty much any model with a mean relational duration under about 50-75 or so. 

I have a bunch of different experiments going right now, and I kind of got it all into a mess.  Let me spend a little time to clean up and document a few examples.  I do need to spend most of tomorrow on another task, though.  I'll try to do it in parallel.

**Comment by sgoodreau** [comment]
OK some of my overnight experiments turned out successful with just using `EGMME.MCMC.burnin.min` and  `EGMME.MCMC.burnin.max`, so I'm feeling more hope than yesterday. As mentioned, I have another task to turn my full focus to, but will circle back to this tomorrow for a bit, and then perhaps over the weekend, or definitely by Monday.

In the meanwhile, I do see two other parameters that were mentioned in the materials above `(discordance_fraction` and `~strat(...) + discord + sparse)`.  I'd also like to give them a try, so if you have any memories to guide those beyond what appears in the thread above, LMK.

Also going to reach out to Deven, although I suspect he's also in the "Approximation Forever" camp :-)

And eventually to Pavel, but only after I plumb the depths of existing material.

**Comment by sgoodreau** [comment]
@smjenness an update.

I stepped back to get much more systematic in my explorations on the tergm side. Have hit some bumps but then made progress again after each one.  I recruited @dth2 for guidance along the way and he's been invaluable.  Of course progress is slow since experiments can take so long, and some of them end with errors. Regardless, there is so much to explore here, and so much room to produce helpful material for both new and seasoned users. Actual write-up on all this should be reasonably straightforward, once I actually understand many of the moving parts.

In terms of the ergm-side write-up, I'm assuming that all will be much easier given that there are so many existing examples and knowledge within the group. It is on my to-do list as well.

All in all we will should have some good materials in time for this year's SISMID, and a nice start towards something larger (i.e. for a larger vignette and/or book chapter).

**Comment by sgoodreau** [comment]
@smjenness quesiton for you:

When you want to set a max degree constraint in your work, do you typically use the term `degrange(from=x)` with target stat of 0?  Or the formal constraint `~bd(maxout=x)`?

I seem to remember us mostly using the former in SISMID, probably because it uses familar syntax so is one less thing to teach.  But as I fit various models with full stergm, I find that this sometimes struggles when the corresponding model with `bd` constraint does better.  At least some of the time this seems to be because SAN lets through a node above the constraint, and then a coefficient gets estimated and the model is very unstable.

I wonder how much this has contributed to past issues with stergm fitting more generally?

Also - since I'll likely want to use `bd` in this fitting tutorial--do we introduce this anywhere in the existing materials?

**Comment by smjenness** [comment]
Always use degrange with target stat of 0, just for simplicity with defining formula in one place. Never had an issue with fitting with it fit this way.

**Comment by sgoodreau** [comment]
Adding in example code from Deven for a detailed ergm fit, so it's here in the record:

```
fit_main.het <- netest(nw_main.het,
                       formation = model_main.het,
                       coef.form = c(-Inf, -Inf),
                       target.stats = netstats_main.het,
                       coef.diss = netstats$main.het$diss.byage,
                       constraints = ~bd(maxout = 2) + blocks(attr = ~sex, levels2 = diag(TRUE, 2)),
                       verbose = TRUE,
                       set.control.ergm = list(SAN.prop = ~strat(attr = ~paste(age.grp, race), pmat=dem.matrix.mh) + sparse,
                                               MCMC.prop = ~strat(attr = ~paste(age.grp, race, het), empirical = TRUE) + sparse,
                                      init.method = "MPLE",
                                      MCMLE.effectiveSize=NULL,
                                      MCMC.burnin=1e6,
                                      MCMC.interval=1e5,
                                      MCMC.samplesize=10000,
                                      init.MPLE.samplesize = 2e8,
                                      MPLE.constraints.ignore = TRUE,
                                      parallel = 4,
                                      SAN.nsteps = 2e5))
```

**Comment by sgoodreau** [comment]
@smjenness Another question, as I keep gathering info on the range of useful practices:

How often do you fit a model on a small population, then manually adjust the edges coefficient to a larger population, and then do your simulations on that? I remember this used to be a thing, but I'm not sure if it still is or not.

**Comment by sgoodreau** [comment]
adding Deven's `tergm` advice for the record:

For the models I need for CAMP projects the edapprox is sufficient for my needs since the model is there for driving simulations rather than for inference. So I mostly just use the edapprox. I have fit a few full stergms and the most helpful terms for me have been the ones you listed. The defaults for the burnin were too small for really large networks so I would generally increase them so they were at least the same order of magnitude as my network size. I don't know if that is a "good" rule of thumb, but it is what I did. What was really helpful was  the MCMC.prop = ~strat(...) + sparse. The strat(...) allowed me to pass stegrm a matrix of sampling weights based on the number of ties between specific demographic groups. This made it possible to avoid repeatedly sampling very low density or empty regions of the mixing matrix. I would use the SAN.prop argument the same way and increase SAN.maxit to 20 so I would have a starting network that was very close to my target stats.

**Comment by sgoodreau** [comment]
@krivit I could use some clarity on one issue with tergm fitting using EGMME and target stats/durations. At this stage I am looking for just **your first impressions** when I describe the issue, without making you dive in deep to try to find solutions, since I know that takes time. Thank you in advance!

Also tagging @smjenness, @AdrienLeGuillou, @dth2 and @martinamorris

The short version: I am finally writing up a guide for folks who use these tools within the context of `EpiModel`, and want to make sure I understand as much as I can. I am struggling with one thing in different forms. I have a model for which the simulations are systematically slightly off the targets (e.g. roughly 390 edges instead of 400).  Not a huge bias, but consistent, and with a Z value of -9. I have tweaked various fitting parameters to improve this, but things still seem to always go to that same place. The consistency of this feels informative. 

**What might cause this, and is there a simple fix?** My leading hypothesis is that perhaps there is some sort of tolerance parameter I’m not aware of, and since the approximation that it starts with is even lower, this is where it always gets to before stopping estimation? If so, can I easily fix that?

Full story: 

The code is laid out below. I don’t think you need to run it to answer the question to give your initial impressions of possible sources.

Network is just 1000 nodes. Model has `edges + concurrent + absdiff age`.  Durations are quite short, averaging 10 time steps. Yes, I realize one solution is to re-define the time steps to be shorter and thus more numerous, and I will present that to folks as an option.  But I am aiming for a more robust set of options, here and overall. And also just to understand the algorithm better.

Using just the approximation does OK but is clearly insufficient; edges is around 385 instead of 400.

Fitting with defaults in `stergm` leads to an error.

Fixing burnin with `EGMME.MCMC.burnin.min = EGMME.MCMC.burnin.max = 1e4` causes the model to fit in roughly half an hour.  And the simulations then average roughly 390 edges. 

Expanding to a burnin min/max of 1e5 causes it to take much longer (obviously), but then yield roughly the same fit with the same simulation outputs.

Tweaking some of the other things that I’ve seen discussed in the past (`SAN.maxit = 20`, using the `sparse` hint) don’t seem to improve further either.

Other suggests I’ve seen don’t seem relevant, in particular `~strat`, since there aren’t a huge number of structural zeros.  I guess accounting for the age mixing here could possibly help, although I think that would mostly just sped up estimation slightly, not fix the issue of always converging to something slightly off.

The code is using `EpiModel` functions; I can rewrite if it helps using plain `stergm` code, although again I don’t think you need to run the code to come up with a first impression of whether there’s something obvious here I don’t know about.

Thank you!!!

```
## Set up

library(EpiModel)
rm(list=ls())
set.seed(1)

nsize <- 1000
nw <- network_initialize(n = nsize)
nw <- set_vertex_attribute(nw, "age", sample(18:65, nsize, replace = TRUE))

coef_diss_10 <- dissolution_coefs(~offset(edges), 10, 0)
formation_2 <- ~edges + concurrent + absdiff("age")
target_stats_2 <- c(400, 100, 2000)

## Fit 1: approximation: edges average ~385

set.seed(1)
system.time(
  est_2_10a <- netest(nw, formation = formation_2, verbose = FALSE,
                target.stats = target_stats_2, coef.diss = coef_diss_10)
            )
summary(est_2_10a)
netdx_2_10a <- netdx(est_2_10a, nsims = 10, nsteps = 1000) 
netdx_stat_2_10a <- netdx(est_2_10a, dynamic = FALSE, nsims = 1e4) 
netdx_2_10a
netdx_stat_2_10a
plot(netdx_2_10a)
plot(netdx_stat_2_10a)

## Fit 2: defaults: Errors out after ~2 hours.

# system.time(
#   est_2_10d <- netest(nw, formation = formation_2, verbose = FALSE, edapprox = FALSE,
#                 target.stats = target_stats_2, coef.diss = coef_diss_10)
#             )
# summary(est_2_10d)
# 
# netdx_2_10d <- netdx(est_2_10d, nsims = 10, nsteps = 1000) 
# netdx_2_10d
# plot(netdx_2_10d)

## Fit 3: fix burnin max/min at 1e4: Edges average ~390

system.time(
  est_2_10b4 <- netest(nw, formation = formation_2, verbose = FALSE, edapprox = FALSE,
                target.stats = target_stats_2, coef.diss = coef_diss_10,
                set.control.tergm = control.tergm(EGMME.MCMC.burnin.min = 1e4, 
                                                  EGMME.MCMC.burnin.max = 1e4))
            )
summary(est_2_10b4)
netdx_2_10b4 <- netdx(est_2_10b4, nsims = 10, nsteps = 1000) 
netdx_2_10b4
plot(netdx_2_10b4)

## Additional efforts all head to roughly 390.

**Comment by smjenness** [comment]
Just to clarify, the cross-section ERGM fit for the model (edapprox = TRUE in netest, but dynamic = FALSE in netdx) is perfect (right at 400 edges). So the problem is the edapprox itself, not some kind of tolerance for the ERGM. I think that's what you were saying, but just wanted to confirm. I am playing around with a few STERGM control settings to see what might work best there.

**Comment by smjenness** [comment]
I re-diagnosed your exact model (N=1000, `~edges + concurrent + absdiff("age")`, targets 400/100/2000, mean duration 10, fit with `EGMME.MCMC.burnin.min = .max = 1e4`). The fitted coefficients are correct. The ~390 edges with Z ≈ -9 come entirely from running `netdx()` at its default per-step burn-in, which under-mixes at short duration.

At D=10 and a tenth of the edges dissolving every step, the default per-step burn-in does not give it enough proposals to get back to the formation target before the step ends.

The control is `set.control.tergm` in `netdx` (and the same in `netsim`):

```r
# Your fit, unchanged
est <- netest(nw, formation_2, target.stats = target_stats_2, coef.diss = coef_diss_10,
              edapprox = FALSE,
              set.control.tergm = control.tergm(EGMME.MCMC.burnin.min = 1e4,
                                                EGMME.MCMC.burnin.max = 1e4))

# What you ran: netdx at its DEFAULT per-step burn-in  ->  edges ~390, Z ~ -9
nd_default <- netdx(est, nsims = 10, nsteps = 1000)

# Same fit, per-step diagnostic burn-in pinned  ->  edges ~400, Z ~ 0
nd_mixed <- netdx(est, nsims = 10, nsteps = 1000,
                  set.control.tergm = control.simulate.formula.tergm(
                    MCMC.burnin.min = 1e4, MCMC.burnin.max = 1e4,
                    MCMC.maxchanges = .Machine$integer.max))
```

Edges recovered from the *same* fit, varying only the diagnostic burn-in:

| netdx per-step burn-in | edges (target 400) | Z |
|---|---|---|
| default | 390.7 | -8.3 |
| 1e4 | ~401 | ~0 |
| 1e5 | 400.9 | ~0 |

Practical notes:

- Set the same `set.control.tergm` in `netsim`. The production run steps the network with the same kernel, so without it the epidemic runs on the under-mixed ~390-edge network too. This is not purely cosmetic.
- 1e4 is enough here and is cheap at this size (about 0.5 s of network stepping per 500-step sim). 1e5 buys nothing more; 1e3 is too few and collapses to ~320 edges.
- This is the same per-step burn-in effect that shows up on the edges-approximation side.

**Comment by sgoodreau** [comment]
THANK YOU!

I feel a bit silly not thinking of the `netdx` as the possible source of the problem and solution here.  What can I say, I'd mentally compartmentalized my task into three separate items (tweaking STERGM estimation, tweaking ERGM estimation, tweaking simulation including diagnostics) and somehow never thought to think ahead to ways they interact.

This actually helps build out a really nice narrative for the tutorial.

Happy Steve.

> I re-diagnosed your exact model (N=1000, `~edges + concurrent + absdiff("age")`, targets 400/100/2000, mean duration 10, fit with `EGMME.MCMC.burnin.min = .max = 1e4`). The fitted coefficients are correct. The ~390 edges with Z ≈ -9 come entirely from running `netdx()` at its default per-step burn-in, which under-mixes at short duration.
> 
> At D=10 and a tenth of the edges dissolving every step, the default per-step burn-in does not give it enough proposals to get back to the formation target before the step ends.
> 
> The control is `set.control.tergm` in `netdx` (and the same in `netsim`):
> 
> # Your fit, unchanged
> est <- netest(nw, formation_2, target.stats = target_stats_2, coef.diss = coef_diss_10,
>               edapprox = FALSE,
>               set.control.tergm = control.tergm(EGMME.MCMC.burnin.min = 1e4,
>                                                 EGMME.MCMC.burnin.max = 1e4))
> 
> # What you ran: netdx at its DEFAULT per-step burn-in  ->  edges ~390, Z ~ -9
> nd_default <- netdx(est, nsims = 10, nsteps = 1000)
> 
> # Same fit, per-step diagnostic burn-in pinned  ->  edges ~400, Z ~ 0
> nd_mixed <- netdx(est, nsims = 10, nsteps = 1000,
>                   set.control.tergm = control.simulate.formula.tergm(
>                     MCMC.burnin.min = 1e4, MCMC.burnin.max = 1e4,
>                     MCMC.maxchanges = .Machine$integer.max))
> Edges recovered from the _same_ fit, varying only the diagnostic burn-in:
> 
> netdx per-step burn-in	edges (target 400)	Z
> default	390.7	-8.3
> 1e4	~401	~0
> 1e5	400.9	~0
> Practical notes:
> 
> * Set the same `set.control.tergm` in `netsim`. The production run steps the network with the same kernel, so without it the epidemic runs on the under-mixed ~390-edge network too. This is not purely cosmetic.
> * 1e4 is enough here and is cheap at this size (about 0.5 s of network stepping per 500-step sim). 1e5 buys nothing more; 1e3 is too few and collapses to ~320 edges.
> * This is the same per-step burn-in effect that shows up on the edges-approximation side.

**Comment by sgoodreau** [comment]
@smjenness As I write up the SISMID version, I am assuming that by this point (1) the audience will have heard of SAN and MCMC and the like, but (2) not been told much about them; and (3) we aren't interested in providing a deep-level training in ERGM and TERGM estimation overall. Thus I'm just providing a very brief (2-3 sentence) high-level picture of what is going on, with pointers to the literature and help files for more info. Sound about right? I imagine we might choose to expand on this in other settings, or even here in the future, perhaps as a supplement, but not now.

**Comment by smjenness** [comment]
Agreed

**Comment by sgoodreau** [comment]
@smjenness So I sent that comment, clsoed GitHub, and then promptly went into the zone and wrote up more like a few paragraphs. Oops :-) But now that I have it there, it does feel like it's roughly the right level for people to be able to understand some basic fitting parameters without getting too lost in weeds.  I'll push forward, and we'll triangulate to the right place.

**Comment by sgoodreau** [comment]
@smjenness @AdrienLeGuillou @dth2 @krivit @martinamorris 

Is it your understanding that the concept of "discordance" in both of these uses:

`MCMC.prop = ~ discord`
`MCMC.prop.args = list(discordance_fraction = x)`

refers to the same thing, namely a dyad that is in a toggled state relative to its start at the beginning of the current MCMC?

If so, then can anyone answer these questions:

- What does `MCMC.prop = ~ discord` do, concretely, and when is it useful? Can it be used independently, or only in conjuniction with `discordance_fraction` (i.e. are they different things, or is the latter is the required parameter for the former)?
- What is an exmaple of a scenario in which setting `discordance_fraction` higher its default value of 0.5 might be useful?  Lower?

Thanks.

**Comment by sgoodreau** [comment]
PS I have read the little documentation I can find on these (including?ergmHint); but it's very minimal, and terms are not always defined.

**Comment by dth2** [comment]
I have never used  MCMC.prop.args = list(discordance_fraction = x)so I don’t know what it does.

**Comment by krivit** [comment]
> * What does `MCMC.prop = ~ discord` do, concretely, and when is it useful? Can it be used independently, or only in conjuniction with `discordance_fraction` (i.e. are they different things, or is the latter is the required parameter for the former)?

You know how TNT improves mixing in sparse networks by focusing on edges? Discord does that, except that it focuses on dyads that have changed since the start of the time step (in either direction) rather than edges. The logic is that for a dynamic network, it is the set of changes from the previous network that is "sparse".

Just as TNT needs a tuning parameter (currently hard-coded at 1/2) to control how often to propose an edge as opposed to a dyad in general, discordance TNT needs one to control how often to propose a discordant dyad as opposed to a dyad in general.

> * What is an exmaple of a scenario in which setting `discordance_fraction` higher its default value of 0.5 might be useful?  Lower?

No idea about this. I don't know if we've ever benchmarked it for TNT in the first place, but I have a vague memory of doing the maths and showing that 1/2 was optimal in some sense.

**Comment by sgoodreau** [comment]
@krivit @dth2 Thanks.

OK, that clarifies the definition of ~discord (in the documentation it is simply referred to as relating to "discordant edges" with no explanation of what that means; I only figured out from a mention under `~discordance_fraction` that that likely meant dyads that have changed since the last time step.

I think you implicitly answered this question, but I just want to be 100% clear, given the documentation gaps: is it the case then that `MCMC.prop = ~ discord` is a simple way to set the `discordance_fraction` at ~0.5, and then setting `discordance_fraction` is just an alternate way to set the same parameter, but with an ability to pick an arbitrary value?  Or do they do different things? This was the basic question I meant to ask, but I think you jumped over it because it probably seems so obvious to you.

The context here is that I have taken on the task of trying to produce a tutorial for folks trying to navigate model fitting in `EpiModel`, which is very much needed. But of course the difficulty with trying to document something without much documentation is that there isn't much documentation.  And it's not stuff I know vey well myself.

**Comment by sgoodreau** [comment]
@smjenness 

OK I am at the point of trying to see if I can write up a stand-alone example when using `edapprox` is sufficient but things are complex enough to require the set of `ergm` control parameters you listed above. You mentioned the GlobalMix India example above and gave lots of detail, but not the actual model or target stats or the like.  Plus it seems exceedingly complex for a stand-alone example (e.g. it involves 24 different networks?!)

Similarly, none of the examples linked to throughout this issue provide sufficient detail about the model and target stats to be able to produce a standalone example, at least that I can see. And I don't have a lot of time at this point to do all of the trial and error that I did for tergm.

If you can point me to something that's a bit closer to having sufficient information to reproduce then I can do some exploration to try to get there.  Otherwise I can just start tomorrow by making up something that is on a really large network and has lots of terms, but long relational durations so the approx works OK, and see where that gets me. I'd rather not start with something that is weird because of one very specific reason (like 1-2 of the examples) -- rather, something that is just generally complex.

Worse comes to worst I'll revert to just making a table of fitting parameters on the ergm side, and be done with it for now.

Sorry this hasn't gone the way I think you imagined it would.

**Comment by smjenness** [comment]
honestly, I think we should table the non-edapprox issues for SISMID. After discussing last week, I'm not even sure a tutorial is the right approach here to highlight the key ERGM fitting controls that I was most concerned with (instead a table or other summary would likely be sufficient). Feel free to feature what you have about the approximation, but don't worry about the issues outside the approximation for now.

**Comment by sgoodreau** [comment]
Thanks Sam.  I agree, now that I've seen how the full body of material, a SISMID tutorial doesn't feel like the logical place to start for most of it.

I'll wrap what I have today, with some form of ERGM fitting table in it. You can take a look later (prob Sunday at your house) and we can decide whether and what we should actually go through in NME-II this year. I won't be hurt if we decide on none of it.  

I do think we made good progress on getting started in a bunch of important directions, and got the conversations going.  And your compilation of material from different sources in the comments above is itself a huge step. We just need to make sure to keep this at the top of the list after SISMID and not let it drop until next year. I don't think we have to be too worried; you seem to have a good vision for some material for more advanced users, and I will need to include something for more beginner users in the book.
